lightning-invoice/src/payment.rs

```diff
@@ -188,10 +190,81 @@ where
 	logger: L,
 	event_handler: E,
 	/// Caches the overall attempts at making a payment, which is updated prior to retrying.
-	payment_cache: Mutex<HashMap<PaymentHash, PaymentAttempts<T>>>,
+	payment_cache: Mutex<HashMap<PaymentHash, PaymentInfo<T>>>,
 	retry: Retry,
 }
 
+/// Used by [`InvoicePayerUsingTime::payment_cache`] to track the payments that are either
+/// currently being made, or have outstanding paths that need retrying.
+struct PaymentInfo<T: Time> {
+	attempts: PaymentAttempts<T>,
+	paths: Vec<Vec<PaymentInfoRouteHop>>,
+}
+
+/// Subset of properties that is represented in `RouteHop`, so we only store what we need in
+/// [`InvoicePayerUsingTime::payment_cache`]
+///
+/// [`RouteHop`]: crate::routing::router::RouteHop
+#[derive(Clone, PartialEq, Eq)]
+struct PaymentInfoRouteHop {
+	/// The node_id of the node at this hop.
+	pub pubkey: PublicKey,
+	/// The channel that should be used from the previous hop to reach this node.
+	pub short_channel_id: u64,
+	/// The fee taken on this hop (for paying for the use of the *next* channel in the path).
+	/// For the last hop, this should be the full value of the payment (might be more than
+	/// requested if we had to match htlc_minimum_msat).
+	pub fee_msat: u64,
+}
+
+impl<T: Time> PaymentInfo<T> {
+	fn new() -> Self {
+		PaymentInfo {
+			attempts: PaymentAttempts::new(),
+			paths: vec![vec![]],
+		}
+	}
+}
+
+/// Used to store information about all the HTLCs that are inflight across all payment attempts
+struct AccountForInFlightHtlcs<'a, S: Score> {
```

Comment:
yeah, I'm a little stuck trying to untangle myself with passing the wrapped scorer to `router.find_route`

```rust
let inflight_htlcs = {...};
let scorer = AccountForInFlightHtlcs { scorer: &mut self.scorer.lock(), inflight_htlcs };

let route = self.router.find_route(..., &scorer);
```

which causes an error:

```
note: expected reference `&<<S as Deref>::Target as LockableScore<'_>>::Locked`
               found reference `&AccountForInFlightHtlcs<'_, <<S as Deref>::Target as LockableScore<'_>>::Locked>`
```

The error is expected since our impl currently expects our scorer to be `Deref` and likewise, our router is expecting its scorer to be a `LockableScore`. And that's pretty much where I'm stuck because I'm rather unfamiliar with the type assertions. I tried removing the constraints for `R` and `S`, and just have `S: Score` instead of `S: Deref` but I ended up opening a barrage of errors in tests, so it looked like I was doing something wrong.

Thoughts?

Reply:
Hmm... the problem is that `Router` is expecting the same `Score` implementation as `InvoicePayer`, but we are wrapping it with `AccountForInFlightHtlcs`. Moving the `Score` type parameterization from `Router` to `Router::find_route` and changing the `Router` type parameterization on `InvoicePayer` to simplly `R: Router` will fix this. This is functionally equivalent to moving the `Score` parameterization to `DefaultRouter` as per https://github.com/lightningdevkit/rust-lightning/pull/1643#issuecomment-1209810558.

Will also need to change the above lines to:

```rust
let mut locked_scorer = self.scorer.lock();
let scorer = AccountForInFlightHtlcs { scorer: &mut locked_scorer, inflight_htlcs };
```